Numpy literals

johnc1231 opened the pull request:
This PR adds the ability to use `hl.literal` on numpy arrays. It also changes the behavior of `hl._ndarray` when called on a numpy ndarray so it just goes through literal to avoid creating an IR node for each element of tensor.

johnc1231 commented:
Partially addresses #7122

tpoterba commented on hail/python/hail/expr/types.py:
```diff
@@ -1647,6 +1656,16 @@ def types_match(left, right) -> bool:
+    elif np_dtype == np.float32:
+        return tfloat32
+    elif np_dtype == np.float64:
+        return tfloat64
```
needs a base case that errors

johnc1231 commented:
Lack of test was me being lazy since I moved all the tests in my other ndarray PR. I'll just add one here and address the merge conflict in either direction.

johnc1231 commented:
Agh, mostly unrelated tests are failing because of a bug that I've fixed in a different PR. I'll wait for that to go in (should go in tomorrow), then rebase and dismiss

johnc1231 commented:
Actually closing this for a minute. Realized I'm not respecting row / column major semantics.

johnc1231 commented:
Reopening this. I've addressed your comments, and also made the decision to take the `row_major` argument off of `_ndarray`, since it adds unnecessary complexity that our users don't benefit from since we are reading everything into json anyway.

tpoterba commented on hail/src/main/scala/is/hail/expr/types/encoded/EBaseStruct.scala:
```diff
@@ -86,6 +86,9 @@ final case class EBaseStruct(fields: IndexedSeq[EField], override val required:
         PTupleField(idx, pt)
       }
       PTuple(pFields, required)
+    case t: TNDArray =>
```
I'm a little unsure about this stuff, but good for now.
